cmd/buildctl/build/ocilayout.go

```diff
@@ -25,3 +31,57 @@ func ParseOCILayout(layouts []string) (map[string]content.Store, error) {
 
 	return contentStores, nil
 }
+
+func newIndexedStore(root string) (content.Store, error) {
+	store, err := local.NewStore(root)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := os.Open(path.Join(root, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var idx ocispecs.Index
+	if err := json.NewDecoder(f).Decode(&idx); err != nil {
+		return nil, err
+	}
+
+	return &indexedStore{
+		Store: store,
+		index: idx,
+	}, nil
+}
+
+type indexedStore struct {
+	content.Store
+	index ocispecs.Index
+}
```

Comment:
This is the bit I'm not so sure about. We need some way on the server to work out what tag is present - we could *potentially* just use the `.Walk` available to us, and then fetch all the content, and unpacking it all to look for the right annotations in the manifest. However, with even small content stores, that's unnecessary amounts of content to walk over and scan.

This store overrides the local store, which performs a special filtering operation to provide for when an indexed piece of content appears - probably not ideal, but this works entirely inline with the content store API.

The only other alternative that comes to mind is to create the store as a local directory as well, to be able to read `/index.json` from buildkit, but that feels frustrating, given we only want to read `index.json`, none of the other content.

Unfortunately, what works for the local cache exporter doesn't work here (which is to modify the `--cache-to` parameters using the `index.json` before sending them to buildkit), since we don't know exactly which tags the frontend will request, they don't necessarily need to be sent over in the client request (and if they were, different frontends might take different formats).

Reply:
`Walk()` is really linearly scanning the `index.json` to find all of the tags, eventually stopping when the `fn` matches? In the single use case we have, it just looks for a matching tag?

> However, with even small content stores, that's unnecessary amounts of content to walk over and scan.

And yet, there really is no other way. There is no b-tree format or similar efficient store for `index.json`. It has to be scanner linearly.

The one concern I would have is with using `tag`. I have seen cases where the annotation is used for _just_ the tag, others where it is used for the entire name. And then when you consider canonicalization of the name, it gets messy.

So if my `ref` is `alpine:3.16`, which of the following should it be looking for?

* `3.16`
* `alpine:3.16`
* `docker.io/library/alpine:3.16`

Again, I have seen all of the above used. Where I use oci-layout setups, I am strict about full canonicalization before writing and upon reading, is whether you wrote an image as `alpine:3.16` or `library/alpine:3.16` or `docker.io/library/alpine:3.16`, you got `docker.io/library/alpine:3.16`. Similarly, when reading.

But that is hardly universally agreed.

Reply:
> `Walk()` is really linearly scanning the `index.json` to find all of the tags, eventually stopping when the `fn` matches? In the single use case we have, it just looks for a matching tag?

...yup, pretty much :smile: The code isn't quite completed, it doesn't play well with any of the other filters, so I need to tidy that up. However, overriding `Walk()` on the build client to *only* scan `index.json` instead of needing to traverse the entire content store is an improvement :smile:

 > I have seen cases where the annotation is used for just the tag, others where it is used for the entire name. And then when you consider canonicalization of the name, it gets messy.

I think the best move here is to conform with how we output OCI images today, [using containerd](https://github.com/moby/buildkit/blob/39f6b4e739747c627aea4ea6fc6e323bb2111a92/vendor/github.com/containerd/containerd/images/archive/exporter.go#L137-L138). In this case, `org.opencontainers.image.ref.name` is used to store only the label+digest, while `io.containerd.image.name` is used to store the full name.

It is worth making sure we're happy with the strategy here though: if we decide to support a syntax like `oci-layout://<oci-store>/<image-in-oci-store>:<tag>` instead of just today's `oci-layout://<oci-store>:<tag>`, then we definitely need the long-form name to determine that.

Reply:
Unfortunately, I can't think of anything reasonable that would let us distinguish between `3.16`, `alpine:3.16` and `docker.io/library/alpine:3.16`.

Here's a suggestion:
- We prefer `io.containerd.image.name` over `org.opencontainers.image.ref.name` if both are present
- We attempt to parse the tag name as a full reference (with host) i.e. `docker.io/library/alpine:3.16` using containerd's `reference.Parse`
- Otherwise we assume it's a plain tag, i.e. `3.16` (with a special empty name, so we can detect it later)

With this, in this PR we'd only compare tag values in `index.json` with the tag specified in the oci-store (so if the OCI store contains `x:123` and `y:123`, specifying `oci-layout://<oci-store>:123` could select either.

In a future PR, we should add support for `ci-layout://<oci-store>/<image-in-oci-store>:<tag>`, where we can compare a full name, but I'd rather do that as a follow-up.

Reply:
We could start with the strictest use case, and look at relaxing it over time, i.e. we _only_ support the full canonical name `docker.io/library/alpine:3.16` in `<image-in-oci-store>`. So you **must** do `oci-layout://docker.io/library/alpine:3.16`. We only would ever support the entire canonical name in the store's `index.json`. It does make it easier for us to parse.

One step more flexible, which is almost as easy to handle, is to support anything valid in the `oci-layout://` URI (it sort of is a URI), but canonical names only in the store, i.e. we relax it in `oci-layout://`. So whatever we get in `<image-in-oci-store>` is canonicalized, but it must be valid for that parsing. Just like `docker run` can handle any of these:

* `alpine` (canonicalized to `docker.io/library/alpine:latest`)
* `alpine:3.16` (canonicalized to `docker.io/library/alpine:3.16`)
* `library/alpine:3.16` (canonicalized to `docker.io/library/alpine:3.16`)
* `docker.io/library/alpine:3.16` (already canonicalized)

But just a tag would not be accepted: `latest`, as we cannot tell what that is.

That is the way I have done most things that read/write from an OCI layout, under the assumption that the layout would include multiple _images_, not just multiple _tags_ for a single image, forcing me to do that.

Thoughts?